docs/posts/lms-for-data/index.qmd

```diff
@@ -0,0 +1,695 @@
+    bill_depth_mm: float,  # <1>
+    flipper_length_mm: float,  # <1>
+    body_mass_g: float,  # <1>
+) -> str:  # <1>
+    # <2>
+    instructions = f"""Provide a whimsical poem that rhymes for a penguin.
+
+    You have the following information about the penguins:
+        species {species}
+        island of {island}
+        bill length of {bill_length_mm} mm
+        bill depth of {bill_depth_mm} mm
+        flipper length of {flipper_length_mm} mm
+        body mass of {body_mass_g} g.
+
+    You must reference the penguin's size in addition to its species and island.
+    """  # <2>
+
+    poem = marvin.generate(  # <3>
+        n=1,  # <3>
+        instructions=instructions,  # <3>
+    )  # <3>
+
+    return poem[0]  # <4>
+```
+
+1. Define a scalar Python UDF to generate a poem from penguin data
+2. Augment the LM's prompt with the penguin data
+3. Use Marvin to generate a poem for the penguin data
+4. Return the generated poem
+
+And apply that UDF to our penguins table:
+
+```{python}
+t = (
+    t.mutate(  # <1>
+        poem=penguin_poem(  # <1>
+            t.species,  # <1>
+            t.island,  # <1>
+            t.bill_length_mm,  # <1>
+            t.bill_depth_mm,  # <1>
+            t.flipper_length_mm,  # <1>
+            t.body_mass_g,  # <1>
+        )  # <1>
+    )
+    .relocate("species", "island", "poem")  # <2>
+    .cache()  # <3>
+)
+t
+```
+
+1. Apply the UDF by mutating the table with the function, using other columns as
+ input
+2. Rearrange the columns we care about to the front
+3. Cache the table to avoid re-running the UDF
+
+:::{.callout-note title="Is this RAG?" collapse="true"}
+Retrieval augmented generation (RAG) is a term for retrieving information,
+augmenting the prompt input to a language model with that information, and using
+that prompt to generate a response with improved accuracy.
+
+In the example above, we've retrieved information from a table (that could be in
+a remote database) and used it to augment the prompt input to a language model.
+The model then generated a response that incorporates the retrieved information,
+improving accuracy.
+
+> I would like to abolish the term RAG and instead just agree that we should
+> always try to provide models with the appropriate context to provide high
+> quality answers. - [Hamel
+> Husain](https://twitter.com/HamelHusain/status/1709740984643596768)
+
+I would consider this RAG, but I also consider RAG to be a silly term.
+:::
+
+Nice! While not particularly useful in this case, the same process can be used
+for generating product descriptions or other practical applications.
+
+## Natural language processing
+
+This includes tasks like:
+
+- sentiment analysis
+- named entity recognition
+- part of speech tagging
+- summarization
+- translation
+- question answering
+
+Each of these tasks can be, to some extent, solved by traditional natural
+language processing (NLP) techniques. However, modern-day LMs can solve these
```

Comment:
depends on what you mean by `solve`. i don't think these problems are solved, but in many cases, AI performance can be on par with or even better than human perf (better than human still doesn't mean it's solved)

Reply:
rewording -- "solved" -> "performed"